File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
       Stack<Character> stack = new Stack<>();
        
        // Iterate through each character in the string.
        for (char c : s.toCharArray()) {
            // If the current character is an opening parenthesis, push it onto the stack.
            if (c == '(' || c == '[' || c == '{') {
                stack.push(c);
            } else {
                // If the current character is a closing parenthesis, check if the stack is empty.
                if (stack.isEmpty()) {
                    return false; // Unmatched closing parenthesis.
                }
                
                // Pop the top element from the stack and check if it matches the current closing parenthesis.
                char top = stack.pop();
                if (c == ')' && top != '(' ||
                    c == ']' && top != '[' ||
                    c == '}' && top != '{') {
                    return false; // Mismatched opening and closing parentheses.
                }
            }
        }
        
        // After processing all characters, the stack should be empty for a valid string.
        return stack.isEmpty(); 
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = deque()

        for i in s:
            if i == '(' or i == "[" or i== "{" :
                stack.append(i)
            else :
                
                    if len(stack) == 0:
                        return False
                    
                    if (stack[-1] == "(" and i != ")") or (stack[-1] == "{" and i != "}") or (stack[-1] == "[" and i != "]") :
                       return False
                    stack.pop()
        return not stack


        